src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -177,125 +199,200 @@ case class WorkflowActor(workflow: WorkflowDescriptor,
+      case Some(x) => Success(x)
+      case _ => Failure(new WdlExpressionException(s"Could not resolve $name as a namespace"))
     }
   }
 
-  def fetchLocallyQualifiedInputs(call: Call): Future[Map[String, WdlValue]] = {
-    // TODO: This assumes that each Call has a parent that's a Workflow, but with scatter-gather that might not be the case 
-    val parentWorkflow = call.parent map { _.asInstanceOf[Workflow] } getOrElse {
-      throw new WdlExpressionException("Expecting 'call' to have a 'workflow' parent.")
+  private def lookupCall(key: ExecutionStoreKey, workflow: Workflow)(name: String): Try[WdlObject] = {
+    workflow.calls find { _.name == name } match {
+      case Some(matchedCall) =>
+        /**
+         * After matching the Call, this determines if the `key` depends on a single shard
+         * of a scatter'd job or if it depends on the whole thing.  Right now, the heuristic
+         * is "If we're both in a scatter block together, then I depend on a shard.  If not,
+         * I depend on the collected value"
+         *
+         * TODO: nested-scatter - this will likely not be sufficient for nested scatters
+         */
+        val index: ExecutionIndex = matchedCall.closestCommonAncestor(key.scope) flatMap {
+          case s: Scatter => key.index
+          case _ => None
+        }
+        fetchCallOutputEntries(findOutputKey(matchedCall, index) getOrElse {
+          throw new WdlExpressionException(s"Could not find a callKey for name '${matchedCall.name}'")
+        })
+      case None => Failure(new WdlExpressionException(s"Could not find a call with name '$name'"))
+    }
+  }
+
+  private def lookupDeclaration(workflow: Workflow)(name: String): Try[WdlValue] = {
+    workflow.declarations find { _.name == name } match {
+      case Some(declaration) => fetchFullyQualifiedName(declaration.fullyQualifiedName)
+      case None => Failure(new WdlExpressionException(s"Could not find a declaration with name '$name'"))
     }
+  }
 
-    object CallInputWdlFunctions extends WdlFunctions {
-      def getFunction(name: String): WdlFunction = {
-        throw new WdlExpressionException("TODO: Some functions may be allowed in this context")
+  private def lookupScatterVariable(callKey: CallKey, workflow: Workflow)(name: String): Try[WdlValue] = {
+    val scatterBlock = callKey.scope.ancestry collect { case s: Scatter if s.item == name => s } headOption
+    val scatterCollection = scatterBlock map { s =>
+      s.collection.evaluate(scatterCollectionLookupFunction(workflow, callKey), new NoFunctions) match {
+        case Success(v: WdlArray) if callKey.index.isDefined =>
+          if (v.value.isDefinedAt(callKey.index.get))
+            Success(v.value(callKey.index.get))
+          else
+            Failure(new WdlExpressionException(s"Index ${callKey.index.get} out of bounds for $name array."))
+        case _ => Failure(new WdlExpressionException(s"$name did not evaluate to a WdlArray"))
```

Comment:
doesn't seem like the right error if `Success(v: WdlArray) if !callKey.index.isDefined`